suite/suite.go

```diff
@@ -0,0 +1,256 @@
+//
+//   - SuiteManager lets you specify groups of files to treat as test data.
+//   - SuiteManager lets you associate TestingPattern callbacks with hunk names by globbing.
+//   - SuiteManager automatically names your tests based on the filename and hunk names.
+//   - SuiteManager automatically rigs up fixture regeneration for you when `-testmark.regen=true`.
+//   - SuiteManager will warn you about any hunks that go unused in a file (helps detect typos!).
+//   - SuiteManager will warn you about any hunk globs that go unmatched in a file.
+//
+// In general, using SuiteManager will help make sure your fixture files and test cases
+// stay aligned as the code evolves: if you remove tests that used some hunks, you'll get notified;
+// if you add tests that expect certain patterns of hunks in some files, but didn't write them yet,
+// you'll get notified; etc.
+//
+// Approximately the only thing SuiteManager can't help you with is if you use whole files for organization,
+// and you remove some files (or add some features or tests that would expect whole new files),
+// and you're using globbing to register filenames, then SuiteManager can't help you infer what's missing.
+type SuiteManager struct {
+	fs fsx.FS
+
+	workset map[string]fileContentExpectations
+}
+
+type fileContentExpectations struct {
+	filename string
+
+	handlers map[HunkGlob]TestingPattern
+}
+
+// TestingPattern is the interface that test code should implement to become
+// usable by SuiteManager as a handler.
+//
+// Examples of TestingPattern that are well-known and easy to use immediately
+// include `testexec.TestingPattern`
+//
+// When implementing this interface:
+// The Run method will do the majority of the work.
+// Other methods on the interface let the SuiteManager know how recursion should work,
+// and provide some information that may be used in logging.
+//
+// Note that it is not necessary to import this package in order to implement this interface.
+// Only symbols from testmark's core package and the go standard library are used.
+type TestingPattern interface {
+	Run(
+		t *testing.T, // The standard testing object, for obvious purposes.
+		filename string, // The filename the subject data was loaded from.  Typically not needed (data comes in via `subject` and, if appropriate, can flow out through `patchAccum`).
+		subject *testmark.DirEnt, // The subject hunk (and enclosing dirent, in case you want to navigate to child hunks).
+		reportUse func(string), // Should be called with the full path of any hunk that's consumed by this test.  Used to detect orphaned hunks that went unused by the whole suite.
+		reportUnrecog func(string), // If this test code owns all child hunks, it may call this to report one that it doesn't recognize.
+		patchAccum *testmark.PatchAccumulator, // If non-nil, means regenerating golden master data is requested instead of testing.
+	) error // Run may return errors or call t.Fatal itself.
```

Comment:
I'm uncertain this error return makes any sense.  I lean towards including an `error` return when not sure, but, so far when implementing this, I've always just had `return nil`.

The labeling performed when you return an error might be helpful, but... everything in that message is available in other ways:
- the filename is already in the t.Run name tree
- the hunk name is also already in the t.Run name tree
- the `tp.Name()` isn't, but... the line number in the t.Log output will point at its guts anyway.

Reply:
I'm fine with it returning an `error`. `testing.T.Run` returns a `bool` and it seems more idiomatic to return an error than a bool.

Reply:
Right, but the callback parameter to `testing.T.Run`, which is the more apt comparison, is `func(*T)` with no return value.

I think I'll still side with you in keeping the `error` return though, just because as a heuristic, almost any interface that doesn't return errors seems to surprise and frustrate me in the long run, much more often than not.  And even if I'm wrong, foisting a "return nil" onto implementers isn't a very big burden.